theano/tensor/opt.py

```diff
@@ -3421,7 +3423,21 @@ def local_fill_sink(node):
     c = node.op(*inputs)
     for model in models:
         c = T.fill(model, c)
-    return [c]
+
+    # The newly created node c doesn't has 'clients',
+    # so this iteration is took place with node.outputs[0]
+    replacements = {node.outputs[0]: c}
+    for client, _ in node.outputs[0].clients:
+        if (hasattr(client, 'op') and
+                isinstance(client.op, T.Elemwise) and
+                not client.op == T.fill):
+            # import ipdb; ipdb.set_trace()
+            r = local_fill_sink.transform(client)
```

Comment:
Oh, I think I understand. Here, one input of `client` is still `node.outputs[0]`, not `c`. So, inside that call to `local_fill_sing.transform`, its inputs are not `fill` nodes, and `transform` will just return `False`.
What you actually have to do is _rebuild_ a new node instead of `client`, that takes `c` as input instead of `node.outputs[0]`, and call `local_fill_sink.transform` on that "new" client.

For instance:
```python
for client, cl_idx in node.outputs[0].clients:
    if (...):  # same condition
        client_inputs = client.inputs[:]
        # Replace node.outputs[0]  by c in the client inputs
        client_inputs[cl_idx] = c
        new_client = client.node.op(*client_inputs)
        r = local_fill_sink.transform(new_client)
        ...
```
This is not tested, just off the top of my head.
Could you try to make it work, and check how many times that path of the code is actually called?